File: src/turboquant/polar_quant.py
"""PolarQuant compression pipeline for TurboQuant KV cache.

Implements the full PolarQuant pipeline:
  norm extraction → rotation → scalar quantization → packing

Polar decomposition separates magnitude (norm) from direction (unit vector).
The norm is stored at full float32 precision, while the direction is rotated
via Walsh-Hadamard transform and then quantized with a Lloyd-Max codebook
optimised for Gaussian data.
"""

from __future__ import annotations

import numpy as np
from dataclasses import dataclass
from numpy.typing import NDArray

from src.turboquant.rotation import rotate, inverse_rotate
from src.turboquant.codebook import get_codebook, quantize_scalar, dequantize_scalar

_SUPPORTED_BITS = frozenset({2, 3, 4})


def _next_power_of_two(n: int) -> int:
    """Return the smallest power of 2 >= *n*."""
    if n <= 0:
        raise ValueError(f"n must be positive, got {n}")
    if n == 1:
        return 1
    return 1 << (n - 1).bit_length()


@dataclass(frozen=True)
class CompressedBlock:
    """Single compressed block of a KV cache vector.

    Attributes:
        indices:    Quantized centroid indices (uint8).
                    Length equals the padded (power-of-2) block dimension.
        norm:       L2 norm of the original block, stored as float32.
        n_bits:     Quantization bit-width (2, 3, or 4).
        block_size: Number of elements in the *original* (unpadded) block.
        seed:       Rotation seed used for this block.
    """

    indices: NDArray[np.uint8]
    norm: np.float32
    n_bits: int
    block_size: int
    seed: int


@dataclass(frozen=True)
class CompressedTensor:
    """Collection of compressed blocks representing a full tensor.

    Attributes:
        blocks:         Immutable tuple of :class:`CompressedBlock` instances.
        original_shape: Shape of the tensor before compression.
        n_bits:         Quantization bit-width.
        block_size:     Nominal block size used during compression.
    """

    blocks: tuple[CompressedBlock, ...]
    original_shape: tuple[int, ...]
    n_bits: int
    block_size: int


# ---------------------------------------------------------------------------
# Block-level compress / decompress
# ---------------------------------------------------------------------------


def polar_quantize_block(
    x: NDArray[np.float64],
    n_bits: int,
    seed: int,
) -> CompressedBlock:
    """Compress a single block using PolarQuant.

    Pipeline:
      1. Extract norm: γ = ||x||₂
      2. Normalize:    x̂ = x / γ   (unit vector — "polar" decomposition)
      3. Pad to next power of 2 for WHT (if needed)
      4. Rotate:       ŷ = rotate(x̂, seed)  — entries ≈ N(0, 1/d)
      5. Scale:        z = ŷ · √d            — entries ≈ N(0, 1)
      6. Quantize:     indices = quantize_scalar(z, codebook)

    Zero vectors are handled gracefully: norm=0 and indices map to the
    centroid nearest 0.

    Parameters
    ----------
    x : 1-D float array
        Block data to compress.
    n_bits : int
        Quantization bit-width (2, 3, or 4).
    seed : int
        RNG seed for the Walsh-Hadamard rotation.

    Returns
    -------
    CompressedBlock
    """
    if n_bits not in _SUPPORTED_BITS:
        raise ValueError(
            f"n_bits must be one of {sorted(_SUPPORTED_BITS)}, got {n_bits}"
        )

    x = np.asarray(x, dtype=np.float64).ravel()
    original_len = len(x)
    if original_len == 0:
        raise ValueError("Cannot compress an empty block")

    codebook = get_codebook(n_bits)
    padded_len = _next_power_of_two(original_len)

    # 1. Extract L2 norm
    norm_val = np.float32(np.linalg.norm(x))

    # 2. Zero-vector fast-path
    if norm_val == 0.0:
        indices = quantize_scalar(np.zeros(padded_len), codebook)
        return CompressedBlock(
            indices=indices,
            norm=norm_val,
            n_bits=n_bits,
            block_size=original_len,
            seed=seed,
        )

    # 3. Normalize to unit vector
    x_hat = x / np.float64(norm_val)

    # 4. Pad to next power of 2 for WHT
    if padded_len != original_len:
        x_hat = np.pad(x_hat, (0, padded_len - original_len))

    # 5. Rotate (Walsh-Hadamard + random sign flip)
    y = rotate(x_hat, seed)

    # 6. Scale: rotated unit vector entries ≈ N(0, 1/d), multiply by √d → N(0, 1)
    z = y * np.sqrt(padded_len)

    # 7. Quantize with Lloyd-Max codebook
    indices = quantize_scalar(z, codebook)

    return CompressedBlock(
        indices=indices,
        norm=norm_val,
        n_bits=n_bits,
        block_size=original_len,
        seed=seed,
    )


def polar_dequantize_block(block: CompressedBlock) -> NDArray[np.float64]:
    """Decompress a single :class:`CompressedBlock`.

    Pipeline:
      1. Dequantize:      z  = dequantize_scalar(indices, codebook)
      2. Unscale:          ŷ  = z / √d
      3. Inverse rotate:   x̂  = inverse_rotate(ŷ, seed)
      4. Trim padding:     x̂  = x̂[:block_size]
      5. Rescale:          x  = x̂ · norm

    Returns
    -------
    NDArray[np.float64]
        Reconstructed block of shape ``(block_size,)``.
    """
    # Zero-vector fast-path
    if block.norm == 0.0:
        return np.zeros(block.block_size, dtype=np.float64)

    codebook = get_codebook(block.n_bits)
    d = len(block.indices)

    # 1. Dequantize
    z = dequantize_scalar(block.indices, codebook)

    # 2. Unscale
    y_hat = z / np.sqrt(d)

    # 3. Inverse rotate
    x_hat = inverse_rotate(y_hat, block.seed)

    # 4. Trim padding
    x_hat = x_hat[: block.block_size]

    # 5. Rescale by norm
    return x_hat * np.float64(block.norm)


# ---------------------------------------------------------------------------
# Tensor-level compress / decompress
# ---------------------------------------------------------------------------


def polar_quantize(
    x: NDArray[np.float64],
    n_bits: int,
    seed: int = 42,
    block_size: int = 128,
) -> CompressedTensor:
    """Compress a full tensor using PolarQuant with blocking.

    The tensor is flattened and split into blocks of ``block_size`` elements.
    Each block is compressed independently with a unique seed
    (``seed_i = seed + block_index``).  The last block may be smaller than
    ``block_size`` and will be padded internally to the next power of 2 for
    the Walsh-Hadamard transform.

    Parameters
    ----------
    x : array
        Tensor of any shape.
    n_bits : int
        Quantization bit-width (2, 3, or 4).
    seed : int
        Base RNG seed.  Block *i* uses ``seed + i``.
    block_size : int
        Number of elements per block (default 128).

    Returns
    -------
    CompressedTensor
    """
    if n_bits not in _SUPPORTED_BITS:
        raise ValueError(
            f"n_bits must be one of {sorted(_SUPPORTED_BITS)}, got {n_bits}"
        )
    if block_size < 1:
        raise ValueError(f"block_size must be >= 1, got {block_size}")

    x = np.asarray(x, dtype=np.float64)
    original_shape = x.shape
    flat = x.ravel()
    n = len(flat)

    n_blocks = (n + block_size - 1) // block_size
    blocks: list[CompressedBlock] = []

    for i in range(n_blocks):
        start = i * block_size
        end = min(start + block_size, n)
        blocks.append(polar_quantize_block(flat[start:end], n_bits, seed + i))

    return CompressedTensor(
        blocks=tuple(blocks),
        original_shape=original_shape,
        n_bits=n_bits,
        block_size=block_size,
    )


def polar_dequantize(compressed: CompressedTensor) -> NDArray[np.float64]:
    """Decompress a :class:`CompressedTensor` back to its original shape.

    Returns
    -------
    NDArray[np.float64]
        Tensor with the same shape as the original input.
    """
    parts = [polar_dequantize_block(b) for b in compressed.blocks]
    flat = np.concatenate(parts)
    return flat.reshape(compressed.original_shape)


# ---------------------------------------------------------------------------
# Utilities
# ---------------------------------------------------------------------------


def compression_ratio(n_bits: int, block_size: int) -> float:
    """Calculate the compression ratio versus float16 storage.

    Original cost:    16 bits per value (float16).
    Compressed cost:  *n_bits* per value + 32-bit float32 norm per block.

    .. math::

        \\text{ratio} = \\frac{16}{n\\_bits + 32 / \\text{block\\_size}}

    Parameters
    ----------
    n_bits : int
        Quantization bit-width.
    block_size : int
        Number of elements per block.

    Returns
    -------
    float
        Compression ratio (e.g. ≈3.76× for turbo4 with block_size=128).
    """
    return 16.0 / (n_bits + 32.0 / block_size)

File: src/turboquant_cpp/polar_quant.cpp
#include "polar_quant.h"
#include "codebook.h"
#include "rotation.h"

#include <algorithm>
#include <cmath>
#include <cstring>
#include <numeric>
#include <stdexcept>
#include <string>

#ifdef __AVX2__
#include <immintrin.h>
#endif

namespace turboquant {

// -----------------------------------------------------------------------
// Thread-local scratch buffers — eliminates malloc/free per block.
// With block_size=128 (padded_len=128), each block would otherwise
// allocate 2 vectors (~1KB).  Over 10,000+ blocks per tensor, this
// saves millions of allocations.
// -----------------------------------------------------------------------
static thread_local std::vector<float> tl_float_buf;
static thread_local std::vector<uint8_t> tl_index_buf;

static float* get_float_scratch(int n) {
    if (static_cast<int>(tl_float_buf.size()) < n) {
        tl_float_buf.resize(n);
    }
    return tl_float_buf.data();
}

static uint8_t* get_index_scratch(int n) {
    if (static_cast<int>(tl_index_buf.size()) < n) {
        tl_index_buf.resize(n);
    }
    return tl_index_buf.data();
}

// -----------------------------------------------------------------------
// Block-level compress
// -----------------------------------------------------------------------

CompressedBlock polar_quantize_block(const float* data, int len,
                                     int n_bits, int seed) {
    if (n_bits != 2 && n_bits != 3 && n_bits != 4) {
        throw std::invalid_argument(
            "n_bits must be 2, 3, or 4, got " + std::to_string(n_bits));
    }
    if (len <= 0) {
        throw std::invalid_argument("Cannot compress an empty block");
    }

    const Codebook& cb = get_codebook(n_bits);
    int padded_len = next_power_of_two(len);

    // 1. Compute L2 norm (AVX2 vectorized)
    float norm_sq = 0.0f;
#ifdef __AVX2__
    {
        __m256 acc = _mm256_setzero_ps();
        int i = 0;
        for (; i + 7 < len; i += 8) {
            __m256 v = _mm256_loadu_ps(&data[i]);
            acc = _mm256_fmadd_ps(v, v, acc);
        }
        // Horizontal sum
        alignas(32) float tmp[8];
        _mm256_store_ps(tmp, acc);
        for (int j = 0; j < 8; ++j) norm_sq += tmp[j];
        for (; i < len; ++i) norm_sq += data[i] * data[i];
    }
#else
    for (int i = 0; i < len; ++i) {
        norm_sq += data[i] * data[i];
    }
#endif
    float norm_val = std::sqrt(norm_sq);

    // 2. Zero-vector fast-path
    if (norm_val == 0.0f) {
        float* scratch = get_float_scratch(padded_len);
        std::memset(scratch, 0, padded_len * sizeof(float));
        std::vector<uint8_t> indices(padded_len);
        quantize_scalar(scratch, indices.data(), padded_len, cb);
        return CompressedBlock{
            std::move(indices), norm_val, n_bits, len, seed};
    }

    // 3. Normalize + pad using scratch buffer (no allocation)
    float* x_hat = get_float_scratch(padded_len);
    float inv_norm = 1.0f / norm_val;

#ifdef __AVX2__
    {
        __m256 inv_v = _mm256_set1_ps(inv_norm);
        int i = 0;
        for (; i + 7 < len; i += 8) {
            __m256 v = _mm256_loadu_ps(&data[i]);
            _mm256_storeu_ps(&x_hat[i], _mm256_mul_ps(v, inv_v));
        }
        for (; i < len; ++i) x_hat[i] = data[i] * inv_norm;
    }
#else
    for (int i = 0; i < len; ++i) {
        x_hat[i] = data[i] * inv_norm;
    }
#endif
    // Zero-pad
    for (int i = len; i < padded_len; ++i) {
        x_hat[i] = 0.0f;
    }

    // 4. Rotate (WHT + random sign flip) — in-place, AVX2 accelerated
    rotate(x_hat, padded_len, seed);

    // 5. Scale: rotated unit vector entries ~ N(0, 1/d), * sqrt(d) -> N(0, 1)
    float scale = std::sqrt(static_cast<float>(padded_len));
#ifdef __AVX2__
    {
        __m256 scale_v = _mm256_set1_ps(scale);
        int i = 0;
        for (; i + 7 < padded_len; i += 8) {
            __m256 v = _mm256_loadu_ps(&x_hat[i]);
            _mm256_storeu_ps(&x_hat[i], _mm256_mul_ps(v, scale_v));
        }
        for (; i < padded_len; ++i) x_hat[i] *= scale;
    }
#else
    for (int i = 0; i < padded_len; ++i) {
        x_hat[i] *= scale;
    }
#endif

    // 6. Quantize with Lloyd-Max codebook
    std::vector<uint8_t> indices(padded_len);
    quantize_scalar(x_hat, indices.data(), padded_len, cb);

    return CompressedBlock{std::move(indices), norm_val, n_bits, len, seed};
}

// -----------------------------------------------------------------------
// Block-level decompress
// -----------------------------------------------------------------------

std::vector<float> polar_dequantize_block(const CompressedBlock& block) {
    // Zero-vector fast-path
    if (block.norm == 0.0f) {
        return std::vector<float>(block.block_size, 0.0f);
    }

    const Codebook& cb = get_codebook(block.n_bits);
    int d = static_cast<int>(block.indices.size());

    // 1. Dequantize into scratch buffer
    float* z = get_float_scratch(d);
    dequantize_scalar(block.indices.data(), z, d, cb);

    // 2. Unscale
    float inv_scale = 1.0f / std::sqrt(static_cast<float>(d));
#ifdef __AVX2__
    {
        __m256 inv_v = _mm256_set1_ps(inv_scale);
        int i = 0;
        for (; i + 7 < d; i += 8) {
            __m256 v = _mm256_loadu_ps(&z[i]);
            _mm256_storeu_ps(&z[i], _mm256_mul_ps(v, inv_v));
        }
        for (; i < d; ++i) z[i] *= inv_scale;
    }
#else
    for (int i = 0; i < d; ++i) {
        z[i] *= inv_scale;
    }
#endif

    // 3. Inverse rotate (in-place, AVX2 accelerated)
    inverse_rotate(z, d, block.seed);

    // 4. Trim padding + rescale by norm
    std::vector<float> result(block.block_size);
    float norm_f = block.norm;
#ifdef __AVX2__
    {
        __m256 norm_v = _mm256_set1_ps(norm_f);
        int i = 0;
        for (; i + 7 < block.block_size; i += 8) {
            __m256 v = _mm256_loadu_ps(&z[i]);
            _mm256_storeu_ps(&result[i], _mm256_mul_ps(v, norm_v));
        }
        for (; i < block.block_size; ++i) result[i] = z[i] * norm_f;
    }
#else
    for (int i = 0; i < block.block_size; ++i) {
        result[i] = z[i] * norm_f;
    }
#endif

    return result;
}

// -----------------------------------------------------------------------
// Tensor-level compress (OpenMP parallel across blocks)
// -----------------------------------------------------------------------

CompressedTensor polar_quantize(const float* data, int total_elements,
                                const int* shape, int ndim,
                                int n_bits, int seed, int block_size) {
    if (n_bits != 2 && n_bits != 3 && n_bits != 4) {
        throw std::invalid_argument(
            "n_bits must be 2, 3, or 4, got " + std::to_string(n_bits));
    }
    if (block_size < 1) {
        throw std::invalid_argument(
            "block_size must be >= 1, got " + std::to_string(block_size));
    }

    int n_blocks = (total_elements + block_size - 1) / block_size;
    std::vector<CompressedBlock> blocks(n_blocks);

    #pragma omp parallel for schedule(dynamic, 4) if(n_blocks > 32)
    for (int i = 0; i < n_blocks; ++i) {
        int start = i * block_size;
        int end = std::min(start + block_size, total_elements);
        blocks[i] = polar_quantize_block(data + start, end - start,
                                          n_bits, seed + i);
    }

    std::vector<int> orig_shape(shape, shape + ndim);

    return CompressedTensor{
        std::move(blocks), std::move(orig_shape), n_bits, block_size};
}

// -----------------------------------------------------------------------
// Tensor-level decompress (OpenMP parallel + pre-allocated output)
// -----------------------------------------------------------------------

std::vector<float> polar_dequantize(const CompressedTensor& ct) {
    int total = 1;
    for (int s : ct.original_shape) total *= s;

    int n_blocks = static_cast<int>(ct.blocks.size());

    // Pre-allocate output and write directly (avoids concat copies)
    std::vector<float> result(total, 0.0f);

    #pragma omp parallel for schedule(dynamic, 4) if(n_blocks > 32)
    for (int i = 0; i < n_blocks; ++i) {
        auto part = polar_dequantize_block(ct.blocks[i]);
        int start = i * ct.block_size;
        int copy_len = static_cast<int>(part.size());
        if (start + copy_len > total) copy_len = total - start;
        std::memcpy(&result[start], part.data(), copy_len * sizeof(float));
    }

    return result;
}

double compression_ratio(int n_bits, int block_size) {
    return 16.0 / (n_bits + 32.0 / block_size);
}

}  // namespace turboquant
